### PR TITLE
benchmark| Potential solution for performance regressions

### LINKED ISSUE #1954: Performance Benchmarks / Integration tests
Pylint needs a collection of performance integration tests (see PyCQA/astroid#519) help further optimization efforts. 

They don't need to be run as part of the standard CI or with just calling `pytest` since they will be slow, but we should have them to make sure optimizations efforts aren't over optimizing specific projects and so we know which pull requests will slow down astroid/pylint.

I'm thinking that we should pick a specific tag of a specific project, and then set up `pytest-benchmark` to `setup` by downloading/extracting the correct zip and running pylint against those files on a developer's machine, and timing how long it takes for pylint to execute and if pylint raises any internal errors.

These tests additionally might be used to look for consistency in output and changes in pylint output messages.

### PR DESCRIPTION
**Pull request by doublethefish**
## Steps

- [x] (na) Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] (na) Add a ChangeLog entry describing what your PR does.
- [x] (na) If it's a new feature or an important bug fix, add a What's New entry in `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.

## Description
These commits generate comparable performance data for a run of benchmarked code.

This is a potential 1st step in establishing performance regression tests. Later work would be to actually do the comparison against the target-branch somehow.

We use pytest-benchmark here, rather than asv, because pytest-benchmark was easier to integrate into the existing tests.

### Use-case
A use-case for this functionality is:

1. Write a good benchmark test:
  - See  `tests/benchmark/test_baseline_benchmarks.py` for an example
2. Run `tox -e benchmark`
  - This generates benchmark .json files in the .tox directory
3. Apply some change that you suspect makes the code faster
4. Re-run `tox -e benchmark`
  - This generates another benchmark .json file, probably `00002-something.json`
5. Compare stats:
  - In `.tox/` run `py.test-benchmark compare 0001 0002` - the numbers are used as a glob to find files that match, use the numbers that refer to the `tox -e benchmark` runs you want to compare.
6. Look for statistically significant differences. You will need to use your common sense here.
7. Profit.

### Baseline benchmark for pylint
The baseline benchmark in this PR attempts to benchmark basic runs:
- with no-files
- a single empty file
- lots of empty files
- a single empty checker
- many empty checkers
- a checker that mimics doing expensive work
- across workers and a single-processed

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :hammer: Refactoring (support of)  |
| ✓  | :scroll: Docs (new performance metrics) |

### CODE REVIEW

File: tests/benchmark/test_baseline_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+    sleep_duration = 0.5  # the time to pretend we're doing work for
+
+    def process_module(self, _astroid):
+        """ Sleeps for `sleep_duration` on each call
+
+        This effectively means each file costs ~`sleep_duration`+framework overhead """
+        time.sleep(self.sleep_duration)
+
+
+class SleepingCheckerLong(BaseChecker):
```
Comment (Pierre-Sassoulas) [inline tests/benchmark/test_baseline_benchmarks.py:49]: It look like it's the same or almost the same than the `SleepingChecker`. I suppose the sleep duration will be higher in `SleepingCheckerLong` later ? Maybe, we could inherit from a `BaseSleepingChecker` and just change some attribute to reduce duplication.

File: tests/benchmark/test_baseline_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+        )
+
+    def test_baseline_lots_of_files_j1(self, benchmark):
+        """ Establish a baseline with only 'master' checker being run in -j1
+
+        We do not register any checkers except the default 'master', so the cost is just
+        that of the system with a lot of files registerd """
+        if benchmark.disabled:
+            benchmark(print, "skipping, only benchmark large file counts")
+            return  # _only_ run this test is profiling
```
Comment (Pierre-Sassoulas) [inline tests/benchmark/test_baseline_benchmarks.py:156]:
```suggestion
            return  # _only_ run this test during profiling
```
Or maybe I did not understand the sentence? It's used elsewhere

File: tests/test_functional.py [hunk 3]
```diff
@@ -78,14 +78,22 @@ def get_tests():
...
+def test_functional(test_file, benchmark):
+    benchmark.group = test_file._directory  # group benchmark results by module
     LintTest = (
         LintModuleOutputUpdate(test_file)
         if UPDATE
         else testutils.LintModuleTest(test_file)
     )
     LintTest.setUp()
-    LintTest._runTest()
+    benchmark_blacklist = (
```
Comment (Pierre-Sassoulas) [inline tests/test_functional.py:89]: Maybe add a comment to explain why those are blacklisted, could be helpful to know if this is still relevant later :)

File: tests/unittest_checker_similar.py [hunk 4]
```diff
@@ -196,3 +200,59 @@ def test_no_args():
...
+        # Add the only checker we care about in this test
+        linter.register_checker(similar.SimilarChecker(linter))
+        linter.enable("similarities")
+
+        fileinfos = list(linter._iterate_file_descrs(self._get_filepaths()))
+        assert len(fileinfos) == self.numfiles
+        benchmark(check_parallel, linter, jobs=5, files=fileinfos, arguments=None)
+
+        assert linter.msg_status == 0, (
+            "Similarties checking is broken when called via check_parallel, so exit "
```
Comment (Pierre-Sassoulas) [inline tests/unittest_checker_similar.py:256]: 👍

File: tests/test_self.py [hunk 5]
```diff
@@ -148,27 +159,32 @@ def test_all(self):
             [join(HERE, "functional", "a", "arguments.py")],
             reporter=MultiReporter(reporters),
             code=2,
+            bench=benchmark,
         )
 
-    def test_no_ext_file(self):
-        self._runtest([join(HERE, "input", "noext")], code=0)
+    def test_no_ext_file(self, benchmark):
+        self._runtest([join(HERE, "input", "noext")], code=0, bench=benchmark)
```
Comment (Pierre-Sassoulas) [inline tests/test_self.py:166]: Nitpick : We could name the variable in `_runtest` the same as the current one so it does `bench=bench` or `benchmark=benchmark` everywhere.

File: tests/benchmark/test_baseline_benchmarks.py [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+        fileinfos = [self.empty_filepath for _ in range(self.lot_of_files)]
+        assert linter.config.jobs == 1
+        assert len(linter._checkers) == 1, "Should have 'master'"
+        benchmark(linter.check, fileinfos)
+        assert linter.msg_status == 0, (
+            "Expected no errors to be thrown: %s"
+            % pprint.pformat(linter.reporter.messages)
+        )
+
+    def test_baseline_lots_of_files_j10(self, benchmark):
```
Comment (Pierre-Sassoulas) [inline tests/benchmark/test_baseline_benchmarks.py:168]: Those tests seems duplicated, could it be possible to create a function that we launch with arguments ?
For example:
```
def assert_baseline_lots_of_files(self, jobs, checker):
    inter.config.jobs = jobs
    ...
    assert linter.msg_status == 0, (
        "Expected no errors to be thrown: %s"
        % pprint.pformat(linter.reporter.messages)
    )
def test_baseline_lots_of_files_j1_empty_checker(self, benchmark):
    self.assert_baseline_lots_of_files(1, empty)

def test_baseline_lots_of_files_j10_empty_checker(self, benchmark):
    self.assert_baseline_lots_of_files(10, empty)

....
```

File: tests/benchmark/test_baseline_benchmarks.py (hunk 1 above)
Comment (Pierre-Sassoulas) [reply tests/benchmark/test_baseline_benchmarks.py:49]: Ha my bad, it's a test, it's supposed to be like this. 😅

**Comment by doublethefish** [comment]
PR #3498 is a slice of this PR.

**Comment by doublethefish** [comment]
Having used this code in anger, this PR has been updated to include the most useful change and I think it should go in.

I have removed some of the noise. The files that have been removed muddied the water:
 - The changes to `tests/test_self.py` and `tests/test_functional.py` were a "lets see what happens" commit and it is very debatable if those changes are useful.
    - This is because of how fast most of the tests are and the fact that we're not aiming for high-performance-computing.
    - There are performance wins to be made, especially in `test_functional.py` but those should be targeted with specific benchmarks.
 - Whilst the changes to `tests/unittest_checker_similar.py` are moderately useful, they can wait until we establish that this is the right way to do benchmarking.

**Pull request merged**